Add `rz_deflate*` functions and save projects in compressed format

DMaroo opened the pull request:
# SQUASH ME

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Currently, rizin and Cutter projects are stored without any compression. With this PR, I have added compression features which will allow to store the project files in a compressed format, thus taking up less memory space.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

The integration tests and the unit tests are passing. The CI is green.

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

Fixes #1417.

DMaroo committed:
Add function definitions for rz_deflate and rz_deflatew

 * TODO: integrate the functions in project.c to save rzdb in compressed
   formats

DMaroo committed:
Add function declaration for rz_deflate and rz_deflatew

XVilka commented on librz/util/zip.c:
```diff
@@ -107,4 +106,83 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+}
+
+/**
+ * @brief deflate uncompressed data to zlib or gzipped, use MAX_WBITS as the window size logarithm.
```
Please use `\brief` Doxygen syntax

XVilka commented on librz/util/zip.c:
```diff
@@ -107,4 +106,83 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+ * @brief deflate uncompressed data to zlib or gzipped, use MAX_WBITS as the window size logarithm.
+ * @see rz_deflatew()
+ */
+RZ_API ut8 *rz_deflate(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
```
Please add `RZ_NONNULL` for `src`, also `rz_return_val_if_fail()`

DMaroo replied:
I have added `rz_return_val_if_fail(...)` only in `rz_deflatew(...)` and `rz_inflatew(...)`, since these both are the functions which are indirectly called by other functions (like `rz_deflate(...)`). Putting checks in all the functions would be redundant. I hope this is fine. Thank you.

XVilka commented on librz/util/zip.c:
```diff
@@ -107,4 +106,83 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+ * \return ptr to compressed
+ */
+RZ_API ut8 *rz_deflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen, int wbits) {
+	int err = 0;
```
Same here

DMaroo committed:
Minor changes according to the review

XVilka commented on librz/include/rz_util/rz_file.h:
```diff
@@ -45,6 +45,8 @@ RZ_API char *rz_file_path_unix_to_local(const char *path);
 RZ_API char *rz_file_binsh(void);
 RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen, int wbits);
 RZ_API ut8 *rz_inflate(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen);
+RZ_API ut8 *rz_deflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen, int wbits);
```
When you add those `RZ_*` markers in the function definition, please add them in header files as well. Would be awesome to have them for the `rz_inflate*()` functions as well, btw.

DMaroo replied:
Oh, I am sorry. My bad. Thank you. :)

XVilka commented on librz/util/zip.c:
```diff
@@ -24,31 +24,33 @@ static const char *gzerr(int n) {
  */
-RZ_API ut8 *rz_inflate_ignore_header(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
+RZ_API ut8 *rz_inflate_ignore_header(RZ_NONNULL const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
 	return rz_inflatew(src, srcLen, srcConsumed, dstLen, -MAX_WBITS);
```
Also `rz_return_val_if_fail()`

DMaroo replied:
But wouldn't that be redundant, since the non-null check is present in `rz_inflatew(...)`?

XVilka replied:
These checks appear only in certain "debug" builds, during the release, packaging, etc they evaluate to nothing. See https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#c

DMaroo replied:
Alright, makes sense. I'll add the checks to make it consistent.

XVilka commented on librz/util/zip.c:
```diff
@@ -107,4 +108,83 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+ * \see rz_deflatew()
+ */
+RZ_API ut8 *rz_deflate(RZ_NONNULL const ut8 *src, int srcLen, int *srcConsumed, int *dstLen) {
+	return rz_deflatew(src, srcLen, srcConsumed, dstLen, MAX_WBITS + 32);
```
And here

DMaroo replied:
But wouldn't that be redundant, since the non-null check is present in `rz_deflatew(...)` (same reasoning as above)?

DMaroo committed:
Add RZ_NONNULL; Reorder declarations

DMaroo committed:
Add rz_return_val_if_fail checks to helper functions

DMaroo commented:
Thank you for the review. I'll now go forward with implementing compression in while saving a project.

ret2libc reviewed (commented):
Hi! Please add unit tests for these two functions. Cc @thestr4ng3r

DMaroo committed:
Fix typo in error message

DMaroo committed:
Add null pointer check for RZ_NONNULL parameter

XVilka commented on librz/util/file.c:
```diff
@@ -1183,6 +1183,32 @@ RZ_API bool rz_file_copy(const char *src, const char *dst) {
 #endif
 }
 
+RZ_API bool rz_file_gzip(const char *src) {
```
When you add new API functions  please add Doxygen as well.

XVilka commented on librz/util/file.c:
```diff
@@ -1183,6 +1183,32 @@ RZ_API bool rz_file_copy(const char *src, const char *dst) {
+	return true;
+#else
+	char *src2 = rz_str_replace(strdup(src), "'", "\'", 1);
+	int rc = rz_sys_cmdf("gzip -f '%s'", src2);
```
The whole function is wrong. Why do you need to call `gzip` if you already have `zlib`? Just use these `inflate`/`deflate` functions.

DMaroo replied:
Yes. @thestr4ng3r informed me to use `rz_deflate` and `rz_inflate` functions. I'll be refactoring it now. I'll squash the commits in the end (to keep the commit history clean).

XVilka commented on librz/util/file.c:
```diff
@@ -1183,6 +1183,32 @@ RZ_API bool rz_file_copy(const char *src, const char *dst) {
+#endif
+}
+
+RZ_API bool rz_file_gunzip(const char *src) {
```
Same here

DMaroo committed:
Add zip/unzip functionality when saving a rizin project

* Add RZ_PROJECT_ERR enum for zip/unzip errors
* Add zip functionality when saving
* Add unzip funtionality when loading

DMaroo committed:
Implement file zip/unzip functions

* Implement `rz_file_zip` and `rz_file_unzip`
* Add corresponding Doxygen documentation

DMaroo commented:
The `rz_deflate` doesn't seem to be working correctly. Please feel free to suggest any improvements.

XVilka commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,40 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+ * \param dst destination file (string containing filename)
+ * \return true, if successful; false otherwise
+ */
+RZ_API bool rz_file_zip(const char *src, const char *dst) {
```
It's not ZIP. ZIP is a different format. I suggest to keep it as `rz_file_deflate()` and `rz_file_inflate()` instead.

XVilka commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,40 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+RZ_API bool rz_file_zip(const char *src, const char *dst) {
+	size_t len, comp_len;
+	char *content = rz_file_slurp(src, &len);
+	char *comp_content = rz_deflate(content, len, NULL, &comp_len);
```
Check for NULL in both cases

XVilka commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,40 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+RZ_API bool rz_file_unzip(const char *src, const char *dst) {
+	size_t len, decomp_len;
+	char *content = rz_file_slurp(src, &len);
+	char *decomp_content = rz_inflate(content, len, NULL, &decomp_len);
```
Same here - check for NULL in both cases.

DMaroo committed:
Change rz_file function names and add nullity checks

* Change `rz_file_zip` to `rz_file_deflate` and `rz_file_unzip` to
  `rz_file_inflate`
* Add checks for null pointers
* Fix return value in `rz_project_load_file_raw`

XVilka commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,53 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+
+	char *comp_content = rz_deflate(content, len, NULL, &comp_len);
+	if (!comp_content) {
+		return false;
```
Don't forget `free(content)` here.

XVilka commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,53 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+	}
+
+	char *decomp_content = rz_inflate(content, len, NULL, &decomp_len);
+	if (!decomp_content) {
```
Don't forget `free(content)` here.

XVilka commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,53 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+ * \param dst destination file (string containing filename)
+ * \return true, if successful; false otherwise
+ */
+RZ_API bool rz_file_deflate(const char *src, const char *dst) {
```
Please mark `src` and `dst` as `RZ_NONNULL` and add `rz_return_val_if_fail(src && dst, false)`

XVilka commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,53 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+ * \param dst destination file (string containing filename)
+ * \return true, if successful; false otherwise
+ */
+RZ_API bool rz_file_inflate(const char *src, const char *dst) {
```
Please mark `src` and `dst` as `RZ_NONNULL` and add `rz_return_val_if_fail(src && dst, false)`

XVilka commented on librz/include/rz_util/rz_file.h:
```diff
@@ -67,6 +69,8 @@ RZ_API char *rz_file_tmpdir(void);
 RZ_API char *rz_file_readlink(const char *path);
 RZ_API bool rz_file_copy(const char *src, const char *dst);
 RZ_API RzList *rz_file_globsearch(const char *globbed_path, int maxdepth);
+RZ_API bool rz_file_deflate(const char *src, const char *dst);
```
Also `RZ_NONNULL` here for both `src` and `dst`.

DMaroo committed:
Add RZ_NONNULL to rz_file function params

* Add RZ_NONNULL and nullity checks for function params of
  rz_file_inflate and rz_file_deflate
* Free `content` if any of those two functions fail midway

DMaroo committed:
Fix compiler warnings

* Fixed parameter types and used casting to remove compiler errors like
  -Wpointer-sign and -Wincompatible-pointer-types

XVilka commented on librz/core/project.c:
```diff
@@ -24,8 +24,9 @@ RZ_API RZ_NONNULL const char *rz_project_err_message(RzProjectErr err) {
 		return "migration failed";
-	case RZ_PROJECT_ERR_UNKNOWN:
-		break;
+	case RZ_PROJECT_ERR_ZIP:
```
Since we do not use ZIP, maybe make it more generic? `RZ_PROJECT_ERR_COMPRESSION_FAILED`? Also it's better to separate compression from decompression errors.

DMaroo committed:
Renamed RZ_PROJECT_ERR enum

DMaroo committed:
Add unit tests for rz_inflate and rz_deflate

* As of now, all rz_inflate tests pass, and all rz_deflates tests fail
* Also, mu_asert_streq cannot be used to test in rz_deflate, because
  deflated string contains null terminators. Maybe could use
mu_assert_memeq

ret2libc commented on librz/core/project.c:
```diff
@@ -24,8 +24,9 @@ RZ_API RZ_NONNULL const char *rz_project_err_message(RzProjectErr err) {
-		break;
+	case RZ_PROJECT_ERR_COMPRESSION_FAILED:
+		return "project file compression failed";
+	case RZ_PROJECT_ERR_UNKNOWN: break;
```
style

ret2libc commented on librz/core/project.c:
```diff
@@ -38,22 +39,33 @@ RZ_API RzProjectErr rz_project_save(RzCore *core, RzProject *prj, const char *fi
 }
 
 RZ_API RzProjectErr rz_project_save_file(RzCore *core, const char *file) {
+	char *tmp_file = rz_file_temp("svprj");
```
I think it's better/safer to use `rz_file_mkstemp`. `rz_file_temp` uses `mktemp`, which is not good and probably we should deprecate this function...

ret2libc commented on librz/core/project.c:
```diff
@@ -63,10 +75,19 @@ RZ_API RzProject *rz_project_load_file_raw(const char *file) {
 	}
-	if (!sdb_text_load(prj, file)) {
+
+	char *tmp_file = rz_file_temp("ldprj");
```
same here. mkstemp.

XVilka commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+#include <rz_util.h>
```
I think the test should be called `test_inflate_deflate.c` not `test_zip`.

XVilka commented on test/unit/meson.build:
```diff
@@ -91,7 +91,8 @@ if get_option('enable_tests')
     'util',
-    'vector'
+    'vector',
+	'zip'
```
Broken indentation.

ret2libc replied:
indentation

XVilka commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+
+bool test_rz_inflate(void) {
+	int len0 = strlen(test_cases[0].inflated);
+	unsigned char *inflated0 = rz_inflate((unsigned char *)test_cases[0].deflated, 287, NULL, NULL);
```
I recommend to check also `inflated0` for NULL with `mu_assert_nonnull()`

XVilka commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+	free(null_terminated0);
+
+	int len1 = strlen(test_cases[1].inflated);
+	unsigned char *inflated1 = rz_inflate((unsigned char *)test_cases[1].deflated, 640, NULL, NULL);
```
Same

XVilka commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+	free(null_terminated1);
+
+	int len2 = strlen(test_cases[2].inflated);
+	unsigned char *inflated2 = rz_inflate((unsigned char *)test_cases[2].deflated, 60, NULL, NULL);
```
Same

XVilka commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+}
+
+bool test_rz_deflate(void) {
+	unsigned char *deflated0 = rz_deflate((unsigned char *)test_cases[0].inflated, strlen(test_cases[0].inflated), NULL, NULL);
```
I recommend to check first for NULL with `mu_assert_nonnull`

XVilka commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+	mu_assert_streq((char *)deflated0, test_cases[0].deflated, "rz_deflate failed for 0");
+	free(deflated0);
+
+	unsigned char *deflated1 = rz_deflate((unsigned char *)test_cases[1].inflated, strlen(test_cases[1].inflated), NULL, NULL);
```
Same

XVilka commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+	mu_assert_streq((char *)deflated1, test_cases[1].deflated, "rz_deflate failed for 1");
+	free(deflated1);
+
+	unsigned char *deflated2 = rz_deflate((unsigned char *)test_cases[2].inflated, strlen(test_cases[2].inflated), NULL, NULL);
```
Same

ret2libc commented on librz/core/project.c:
```diff
@@ -38,22 +39,33 @@ RZ_API RzProjectErr rz_project_save(RzCore *core, RzProject *prj, const char *fi
 		return err;
 	}
-	if (!sdb_text_save(prj, file, true)) {
+	if (!sdb_text_save(prj, tmp_file, true)) {
 		err = RZ_PROJECT_ERR_FILE;
 	}
 	sdb_free(prj);
 	if (err == RZ_PROJECT_ERR_SUCCESS) {
 		rz_config_set(core->config, "prj.file", file);
 	}
+
+	if (err == RZ_PROJECT_ERR_SUCCESS) {
+		if (!rz_file_deflate(tmp_file, file)) {
+			free(tmp_file);
+			return RZ_PROJECT_ERR_COMPRESSION_FAILED;
+		}
+	}
+	free(tmp_file);
+
```
```suggestion
	if (!sdb_text_save(prj, tmp_file, true)) {
		err = RZ_PROJECT_ERR_FILE;
	}
	sdb_free(prj);
	if (err != RZ_PROJECT_ERR_SUCCESS) {
		goto tmp_file_err;
	}
		
	if (!rz_file_deflate(tmp_file, file)) {
		err = RZ_PROJECT_ERR_COMPRESSION_FAILED;
		goto tmp_file_err;
	}
	rz_config_set(core->config, "prj.file", file);
tmp_file_err:
	free(tmp_file);
```

ret2libc commented on librz/core/project.c:
```diff
@@ -63,10 +75,19 @@ RZ_API RzProject *rz_project_load_file_raw(const char *file) {
 		return NULL;
 	}
+	free(tmp_file);
+
```
```suggestion
```

ret2libc commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+};
+
+bool test_rz_inflate(void) {
+	int len0 = strlen(test_cases[0].inflated);
+	unsigned char *inflated0 = rz_inflate((unsigned char *)test_cases[0].deflated, 287, NULL, NULL);
+	char *null_terminated0 = malloc(sizeof(char) * (len0 + 1));
+	strncpy(null_terminated0, (char *)inflated0, len0);
+	null_terminated0[len0] = '\0';
+	free(inflated0);
+	mu_assert_streq(null_terminated0, test_cases[0].inflated, "rz_inflate failed for test 0");
+	free(null_terminated0);
```
can we put this in a loop and just iterate over test_cases?

ret2libc commented on test/unit/test_zip.c:
```diff
@@ -0,0 +1,67 @@
+}
+
+bool test_rz_deflate(void) {
+	unsigned char *deflated0 = rz_deflate((unsigned char *)test_cases[0].inflated, strlen(test_cases[0].inflated), NULL, NULL);
+	mu_assert_streq((char *)deflated0, test_cases[0].deflated, "rz_deflate failed for 0");
+	free(deflated0);
```
same with loop ?

XVilka commented:
Also please rebase on top of the latest `dev`.

ret2libc commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,61 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+ * \param dst destination file (string containing filename)
+ * \return true, if successful; false otherwise
+ */
+RZ_API bool rz_file_deflate(RZ_NONNULL const char *src, RZ_NONNULL const char *dst) {
```
I think the implementation of this function is good enough as a first implementaiton, but it requires loading the whole file in memory which imo is not ideal. I believe the zlib library has a stream API by default, so you can call inflate/deflate by providing more data as necessary. If that is so, we could use `rz_buf_new_file` to create a `RzBuffer` from the `src` file and then have a `rz_deflate_buf` function that uses `rz_buf_read` as necessary to read data from the buffer. In this way we don't need to load the whole file in memory. What do you think?

DMaroo replied:
It doesn't work as expected, as of now. I will first need to fix that. After that, I can definitely work on using streaming behaviour instead of storing the whole file in memory.

DMaroo replied:
I have implemented this in my last commit. Please take a look. However, this breaks `rz_file_inflate`. I don't know of any way to have streaming behaviour in `rz_file_inflate`, and just using the current implementation gives `RZ_ERR_INVALID_CONTENTS` error. However, when I manually decompress (using `gzip`) the file generated by `rz_file_inflate` and open that with rizin, it works perfectly fine.

ret2libc replied:
Mmm I'm not sure it works that way. I think you should add `rz_deflate_buf` and in that function you should call `deflateInit2` only one time and then `deflate` multiple times.

Right not it seems to me you just call `rz_deflate` with multiple chunks, but `rz_deflate` essentially starts from scratch with every call. I hope it's clear what I mean :)

ret2libc replied:
I propose to add the APIs:
```C
RZ_API RzBuffer *rz_inflate_buf(RzBuffer *src);
RZ_API RzBuffer *rz_deflate_buf(RzBuffer *src);
```

ret2libc reviewed (commented):
Nice feature to have!

DMaroo committed:
Simplify test_rz_inflate; Rename test_zip; Add notnull check

* Remove unnecessary string operations when using mu_assert_streq.
  Instead use mu_assert_memeq
* Rename test_zip.c to test_inflate_deflate.c
* Add mu_assert_notnull checks for pointers returned by
  rz_inflate/rz_deflate

DMaroo committed:
Use mkstemp to generate a temporary file

* Add RZ_PROJECT_ERR_MKSTEMP_FAILED for mkstemp errors
* Use rz_file_mkstemp instead of rz_file_temp
* Improve program execution flow using error goto

DMaroo committed:
Use for loop to iterate through test cases

DMaroo committed:
Working implementation of rz_deflate

* Still doesn't pass the tests, since some parameters seem to be
  incorrect, but it works, and generates a file with gzip magic number

DMaroo committed:
Replace expected output

* `rz_deflate` does not store name of the file in the deflated data,
  therefore, when compared with plain gzip, it gives incorrect result
* Comparing with `gzip -n` yields correct results
* test_inflate_deflate passing now

DMaroo committed:
Add support for opening non deflated rzdb project files

* Check whether the project file is deflated. Only inflate if the
  project file is deflated
* Makes the change reverse compatible with older project files

DMaroo committed:
Add documentation for rz_file_is_deflated

DMaroo committed:
Fix error code checking for mkstemp; Fix wbit values

ret2libc commented on librz/core/project.c:
```diff
@@ -38,22 +42,40 @@ RZ_API RzProjectErr rz_project_save(RzCore *core, RzProject *prj, const char *fi
 
 RZ_API RzProjectErr rz_project_save_file(RzCore *core, const char *file) {
+	char *tmp_file;
+	int err_mkstemp = rz_file_mkstemp("svprj", &tmp_file);
```
please note that in theory the security guarantees of mkstemp come from the fact that you use the file descriptor returned by it, that is `err_mkstemp` in this case. However, we would need to change few APIs to actually use directly the descriptor instead of taking a file, so... Please close at least the file descriptor returned by this function.

ret2libc commented on librz/core/project.c:
```diff
@@ -63,10 +85,30 @@ RZ_API RzProject *rz_project_load_file_raw(const char *file) {
-	if (!sdb_text_load(prj, file)) {
+
+	char *tmp_file;
+	int err_mkstemp = rz_file_mkstemp("ldprj", &tmp_file);
```
same, at least close the file descriptor.

DMaroo committed:
Delete temporary file aftr project is saved/loaded

* Helpful when opening a huge project multiple times. If we don't delete
  a temp file, a new redundant temp copy is generated for every
  save/load

ret2libc commented on librz/include/rz_project.h:
```diff
@@ -24,6 +24,8 @@ typedef enum rz_project_err {
 	RZ_PROJECT_ERR_INVALID_CONTENTS,
 	RZ_PROJECT_ERR_MIGRATION_FAILED,
+	RZ_PROJECT_ERR_COMPRESSION_FAILED,
+	RZ_PROJECT_ERR_MKSTEMP_FAILED,
```
I don't think it makes too much sense to go that detailed in RzProjectErr about the reason of the failing... Whether we use mkstemp or something else is pretty much an internal detail. Right @thestr4ng3r ?

XVilka replied:
Yes, I suggest to remove MKSTEMP error

DMaroo committed:
Close file (descriptor) returned by rz_file_mkstemp

ret2libc commented on librz/core/project.c:
```diff
@@ -38,22 +42,43 @@ RZ_API RzProjectErr rz_project_save(RzCore *core, RzProject *prj, const char *fi
+	int mkstemp_fd = rz_file_mkstemp("svprj", &tmp_file);
+
+	if (mkstemp_fd == -1 || !tmp_file) {
+		close(mkstemp_fd);
```
if you have to do this both in the if and outside, just put it before the `if` one time.

ret2libc commented on librz/core/project.c:
```diff
@@ -63,10 +88,37 @@ RZ_API RzProject *rz_project_load_file_raw(const char *file) {
+	int mkstemp_fd = rz_file_mkstemp("ldprj", &tmp_file);
+
+	if (mkstemp_fd == -1 || !tmp_file) {
+		close(mkstemp_fd);
```
same

DMaroo committed:
Deflate by using rz_buf with streaming behavior

* No need to store the whole file in memory while deflatting
* TODO: Make rz_inflate work (not working currently)

XVilka commented on librz/core/project.c:
```diff
@@ -38,22 +42,42 @@ RZ_API RzProjectErr rz_project_save(RzCore *core, RzProject *prj, const char *fi
+	close(mkstemp_fd);
+
+	if (mkstemp_fd == -1 || !tmp_file) {
+		return RZ_PROJECT_ERR_MKSTEMP_FAILED;
```
We could return ` RZ_PROJECT_ERR_FILE` here instead.

DMaroo committed:
Removed unnecessary RzProjectErr enum member: RZ_PROJECT_ERR_MKSTEMP_FAILED

XVilka commented:
Project loading tests still do not work:
```
87/88 cpu_platform_profiles          FAIL            0.48s   exit status 1
88/88 open_analyse_save_load_project FAIL            1.88s   exit status 1


The output from the failed tests:

55/88 project_migrate                FAIL            0.35s   killed by signal 11 SIGSEGV
--- command ---
04:42:21 /home/runner/work/rizin/rizin/build/test/unit/test_project_migrate
-------

87/88 cpu_platform_profiles          FAIL            0.48s   exit status 1
--- command ---
04:42:23 /home/runner/work/rizin/rizin/build/test/integration/test_cpu_platform_profiles
--- stdout ---
test_cpu_profiles ERR
[XX] Fail at line 53: project load err: expected 0, got 5.

test_platform_profiles ERR
[XX] Fail at line 104: project load err: expected 0, got 5.

--- stderr ---
Cannot find base type "uid_t"
Cannot find base type "pid_t"
Cannot find base type "uid_t"
Cannot find base type "uid_t"
Cannot find base type "gid_t"
Cannot find base type "pid_t"
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
-------

88/88 open_analyse_save_load_project FAIL            1.88s   exit status 1
--- command ---
04:42:23 /home/runner/work/rizin/rizin/build/test/integration/test_open_analyse_save_load_project
--- stdout ---
test_open_analyse_save ERR
[XX] Fail at line 58: project load err: expected 0, got 5.

--- stderr ---
[ ] Analyze function calls (aac)
[
[x] Analyze function calls (aac)
[ ] Analyze len bytes of instructions for references (aar)
[
[x] Analyze len bytes of instructions for references (aar)
[ ] Check for classes
[
[x] Check for classes
[ ] Type matching analysis for all functions (aaft)
[
[x] Type matching analysis for all functions (aaft)
[ ] Propagate noreturn information
[
[x] Propagate noreturn information
[ ] Use -AA or aaaa to perform additional experimental analysis.
[
[x] Use -AA or aaaa to perform additional experimental analysis.
functions count = 146
-------


Summary of Failures:

55/88 project_migrate                FAIL            0.35s   killed by signal 11 SIGSEGV
87/88 cpu_platform_profiles          FAIL            0.48s   exit status 1
88/88 open_analyse_save_load_project FAIL            1.88s   exit status 1


Ok:                 85  
Expected Fail:      0   
Fail:               3   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   
```

DMaroo commented:
Yes, I am aware that the tests are failing. I will have a look at that once I finish implementing `rz_file_inflate`.

ret2libc commented on librz/util/file.c:
```diff
@@ -1257,3 +1257,92 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+
+	if (!(src_buf && dst_buf)) {
+		free(src_buf);
+		free(dst_buf);
```
```suggestion
		rz_buf_free(src_buf);
		rz_buf_free(dst_buf);
```

DMaroo committed:
Check whether file exists when loading the project from a raw file

* Without the check, rizin exits abruptly with a segfaul (this is maybe
  a regression bug)
* Adding a check solves the problem

DMaroo committed:
Implement rz_deflatew_buf and rz_deflate_buf

* Deflate an RzBuffer, without storing all the contents in program memory

DMaroo committed:
Use rz_deflate_buf in rz_file_deflate

* Chose the block size value to be 256 KB (recommended by zlib). Could be changed later if the
  compression is found ineffective.

DMaroo commented:
A few questions:
 1. `rz_deflate` and `rz_deflatew` are no longer being used anywhere. Should I remove them, or let them be for future convenience?
 2. Should I write any tests for `rz_buf_deflate`?

DMaroo committed:
Implement `rz_inflatew_buf` and `rz_inflate_buf`

* Inflate RzBuffer without storing all the contents in program memory
* Stream the data and inflate accordingly

DMaroo committed:
Use rz_inflate_buf in rz_file_inflate

* Use the (better) rz_inflate_buf in rz_file_inflate
* Currently, the implementation requires ~9 MB memory usage while
  inflating a file. This can be reduced (if necessary) at the expense of
  iterations through the deflated file
* It would also be nice to have a better way to inflate and not waste
  this much memory (most of which is just to accomodate the special
  case of maximum compression ratio)
* Minor changes to rz_file_is_deflated doxygen documentation

DMaroo commented:
I have now done the majority of work for this PR. What remains is only cleaning out unused stuff (if required) and making the CI green. I am now converting this draft to a ready for review PR.

DMaroo commented on librz/util/zip.c:
```diff
@@ -266,3 +266,79 @@ RZ_API bool rz_deflatew_buf(RZ_NONNULL RzBuffer *src, RZ_NONNULL RzBuffer *dst,
+		return false;
+	}
+
+	int comp_factor = 1032; // maximum compression ratio
```
As of now, I just prepare for the highest possible compression ratio and assign memory accordingly, but this forces us not choose a large enough `block_size`. So, we either need to go with a small `block_size` (right now, I am using `1 << 13 = 8 KB` (https://github.com/rizinorg/rizin/pull/1446/commits/b3fa0d5630f65a0deab3d1d08c75e34e845485aa#diff-b6e47a6b965ef83d50289bc0568bb5f82a4e47d5fe654cc6be44a2c85c7b0ce8R1309), which is kind of small for large files, but still amounts to 9 MB memory) and subsequently more number of inflate loops, or go with a larger block size, thus larger memory usage. If there's any better way to handle this, please feel free to suggest.

DMaroo commented:
Also, I haven't profiled this. So I am not sure what is the performance effect on saving huge projects. It would be nice if someone could profile this.

ret2libc commented on librz/util/zip.c:
```diff
@@ -107,4 +110,235 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+ * \param wbits the size of the history buffer (or “window size”), and what header and trailer format is expected.
+ * \return true if successful; false otherwise
+ */
+RZ_API bool rz_deflatew_buf(RZ_NONNULL RzBuffer *src, RZ_NONNULL RzBuffer *dst, ut64 block_size, ut8 *src_consumed, ut8 *dst_len, int wbits) {
```
no need for dst_len, because RzBuffer already has the method rz_buf_size to compute the size.

DMaroo replied:
Wouldn't it be consistent to have this parameter (compared to `rz_deflate`)? Just asking.

ret2libc commented on librz/util/zip.c:
```diff
@@ -107,4 +110,235 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+ * \return true if successful; false otherwise
+ */
+RZ_API bool rz_deflatew_buf(RZ_NONNULL RzBuffer *src, RZ_NONNULL RzBuffer *dst, ut64 block_size, ut8 *src_consumed, ut8 *dst_len, int wbits) {
+	rz_return_val_if_fail(src && dst, false);
```
add:
```
rz_return_val_if_fail(block_size != 0, false);
```

ret2libc reviewed (commented):
It would be great to have unit tests for rz_inflate/deflate_buf functions, yes!

DMaroo commented:
@ret2libc Would adding tests for `rz_file_inflate` and `rz_file_deflate` suffice (since they are only a wrapper around `rz_inflate_buf` and `rz_deflate_buf`, and it would be easier to write and manage tests for `rz_file*` functions, I think)?

DMaroo committed:
Remove a bug which deleted the original file instead of tmp_file

DMaroo committed:
Remove unnecessary dst_len parameter for rz_*flate_buf functions

* Also, add rz_return_val_if_fail check instead of an if statement

DMaroo commented:
I am unable to figure out why the `project_migrate` test is failing.

DMaroo committed:
Add unit tests for `rz_*flate_buf` functions

DMaroo committed:
Change wbits values; Truncate dst buffer

* Changed values of wbits value and memLevel (tests passing)
* Truncate dst buffer before returning from `rz_deflate_buf` to avoid
  trailing garbage

DMaroo committed:
Modify according to clang-format

DMaroo committed:
Use `file` as parameter to `rz_project_save` instead of `tmp_file`

* Stores correct project metadata, and works when moving projects and binaries

ret2libc commented on librz/core/project.c:
```diff
@@ -63,10 +85,41 @@ RZ_API RzProject *rz_project_load_file_raw(const char *file) {
+			prj = NULL;
+		}
+		goto return_goto;
+	}
```
this code is dupped. Rewrite the code so that you just have to switch the argument of `sdb_text_load` instead of rewriting the whole piece of code.

ret2libc commented on librz/util/zip.c:
```diff
@@ -107,4 +109,223 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+ * \param wbits the size of the history buffer (or “window size”), and what header and trailer format is expected.
+ * \return ptr to compressed
+ */
+RZ_API ut8 *rz_deflatew(RZ_NONNULL const ut8 *src, int srcLen, int *srcConsumed, int *dstLen, int wbits) {
```
remove this and rz_inflate and rz_inflatew and rz_deflate. Let's just use the rz_(de|in)flate_buf versions which do not require to read the whole file in mem.

DMaroo replied:
That'll require a few changes in `analysis/flirt.c`, `core/cmd_print.c`, `io/p/io_gzip.c` and `util/file.c` in the `librz` (al the places where `rz_inflate` is being used).

ret2libc replied:
Didn't you say that some functions were never used? What are those functions never used?

XVilka replied:
@ret2libc @DMaroo wait. Lets keep them. FLIRT is tricky to test, and it *does require* wbits option. It has the raw data without the header.

XVilka replied:
Ah, nevermind, they indeed seem to be unused:
```
[i] ℤ rg "rz_inflatew"                                                                                                                                                                                                            17:33:25 
librz/util/zip.c
28: * @see rz_inflatew()
31:     return rz_inflatew(src, srcLen, srcConsumed, dstLen, MAX_WBITS + 32);
36: * @see rz_inflatew()
39:     return rz_inflatew(src, srcLen, srcConsumed, dstLen, -MAX_WBITS);
51:RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen, int wbits) {

librz/include/rz_util/rz_file.h
46:RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLen, int wbits);
rizin on  dev [$] 
[i] ℤ rg "rz_deflatew"   
```
Kill them.

XVilka replied:
The `rz_inflate` is being used:
```
[i] ℤ rg "rz_inflate("                                                                                                                                                                                                             13:17:35 
librz/util/file.c
502:    ut8 *out = rz_inflate(in, (int)sz, NULL, outlen);

librz/io/p/io_gzip.c
169:            mal->buf = rz_inflate(data, (int)len, NULL, size);

librz/core/cmd_print.c
6183:                           out = rz_inflate(block, core->blocksize, &inConsumed, &outlen);
6191:                           out = rz_inflate(block, core->blocksize, NULL, &outlen);
6198:                           out = rz_inflate(block, core->blocksize, NULL, &outlen);

librz/analysis/flirt.c
1395:                   if (!(decompressed_buf = rz_inflate(buf, size, NULL, &decompressed_size))) {
```

ret2libc commented on test/unit/test_inflate_deflate.c:
```diff
@@ -0,0 +1,78 @@
+	{ "1234567890abcdefghijklmnopqrstuvwxyz\n", "\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\x03\x33\x34\x32\x36\x31\x35\x33\xb7\xb0\x34\x48\x4c\x4a\x4e\x49\x4d\x4b\xcf\xc8\xcc\xca\xce\xc9\xcd\xcb\x2f\x28\x2c\x2a\x2e\x29\x2d\x2b\xaf\xa8\xac\xe2\x02\x00\x9f\x79\xdc\xdb\x25\x00\x00\x00" }
+};
+
+int deflated_lengths[] = { 284, 637, 57 };
```
How did you compute these sizes? Can we get them with `sizeof` or `strlen` instead?

DMaroo replied:
`strlen` would stop counting at null bytes. `sizeof` should work fine. I'll try.

DMaroo replied:
I cannot seem to use `sizeof` since it would require an array, and I cannot use `char inflated[]; char deflated[];` in the struct.

ret2libc commented on librz/util/file.c:
```diff
@@ -1252,3 +1252,80 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+	rz_buf_free(src_buf);
+	rz_buf_free(dst_buf);
+	return false;
+}
```
remove this and place `error_route` above. Add a `bool res` variable which is set to false before the goto. This way we don't have to duplicate the code and it's less opportunities for us to forget freeing something.

ret2libc commented on librz/util/file.c:
```diff
@@ -1252,3 +1252,80 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+	rz_buf_free(src_buf);
+	rz_buf_free(dst_buf);
+	return false;
+}
```
same

wargio commented:
> A few questions:
> 
> 1. `rz_deflate` and `rz_deflatew` are no longer being used anywhere. Should I remove them, or let them be for future convenience?
> 2. Should I write any tests for `rz_buf_deflate`?

@XVilka we need to test this on flirt signatures, before merging

DMaroo committed:
Remove redundant code by using local variables

DMaroo committed:
Add OS check to modify deflated data accordingly

* Use an array to store inflated lengths
* Use stack allocated arrays, instead of pointers to read-only memory,
  so that we can modify data in `check_os` function
* Current implementation not very memory effcient, but any other
  implementation would be much more complicated and not a huge
  improvement from memory usage POV (after all, this is just a test)

ret2libc commented on librz/util/zip.c:
```diff
@@ -107,4 +109,223 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+		}
+		err = deflate(&stream, flush);
+		if (err < 0) {
+			RZ_LOG_ERROR("deflate error: %d %s\n", err, gzerr(-err));
```
shall we `break` if there's an error?

DMaroo replied:
Yes, I think we should. zlib specifies negative return codes for `deflate`/`inflate` as errors. I can't think of a case where continuing after getting an error would be helpful.

ret2libc commented on librz/util/zip.c:
```diff
@@ -107,4 +109,223 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+		}
+		err = inflate(&stream, flush);
+		if (err < 0) {
+			RZ_LOG_ERROR("inflate error: %d %s\n", err, gzerr(-err));
```
shall we `break`?

ret2libc commented on test/unit/test_inflate_deflate.c:
```diff
@@ -0,0 +1,99 @@
+
+int deflated_lengths[3] = { 284, 637, 57 };
+int inflated_lengths[3];
+
```
just add a `deflated_length` field in the test_cases above.

ret2libc commented on test/unit/test_inflate_deflate.c:
```diff
@@ -0,0 +1,99 @@
+		inflated_lengths[i] = strlen(test_cases[i].inflated);
+	}
+}
+
```
i think there's no need to compute the inflated_lengths, given they are strings. Just use `mu_assert_streq` in the tests and you are automatically testing also the length of the strings.

DMaroo replied:
We will still need the `inflated_lengths` to pass them to `rz_deflate` and `rz_buf_new_with_bytes`. However, I'll change `mu_assert_memeq` to `mu_assert_streq` for the inflated strings.

ret2libc replied:
You can still just use `strlen(test_case[i].inflated_string)` instead.

DMaroo replied:
Okay, but that felt cumbersome to me, which is why I removed it in the first place. I'll update it now, though.

ret2libc reviewed (changes requested):
Just few minor things, but I think code looks good! Good job and thanks :)

DMaroo committed:
Format according to clang-format; Add deflated_length struct member

DMaroo committed:
Return false when inflate/deflate error occurs

DMaroo committed:
Remove `deflated_lengths` array, instead use `strlen`

ret2libc commented on librz/util/zip.c:
```diff
@@ -107,4 +109,231 @@ RZ_API ut8 *rz_inflatew(const ut8 *src, int srcLen, int *srcConsumed, int *dstLe
+		err = inflate(&stream, flush);
+		if (err < 0) {
+			RZ_LOG_ERROR("inflate error: %d %s\n", err, gzerr(-err));
+			inflateEnd(&stream);
```
as in other places, please use `goto`. Otherwise it becomes too easy to forget about stuff that needs to be freed. Do the same in rz_deflatew_buf

DMaroo committed:
Use goto to avoid code duplication

DMaroo committed:
Missing semicolon

XVilka commented on librz/util/file.c:
```diff
@@ -1252,3 +1252,80 @@ RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+RZ_API bool rz_file_is_deflated(RZ_NONNULL const char *src) {
+	rz_return_val_if_fail(src, false);
+
+	unsigned char *header = (unsigned char *)rz_file_slurp_range(src, 0, 3, NULL);
```
Please check the returned length. If `NULL` or less than 3 bytes returned.

XVilka reviewed (changes requested):
Good job! One minor thing and it's good to go.

DMaroo committed:
Check returned header for null and size; Free memory before return

thestr4ng3r reviewed (changes requested):
Compression should be optional and disabled by default. I suggest adding a config var `prj.compress`.

DMaroo commented:
> Compression should be optional and disabled by default. I suggest adding a config var `prj.compress`.

Alright, makes sense. But why disable it by default. Shouldn't it be enabled by default, given that it is not particularly buggy or slow, and makes the files quite smaller?

UPD: I have pushed a commit which does what you suggested (except it decompresses by default, which can be trivially changed later if needed). It works well when changing options, and no duplicates are created (as I understand from my manual testing).

DMaroo committed:
Add optional parameter for whether to decompress while saving or not

* Add a config variable: `prj.decompress` (bool) to denote whether to
  decompress while saving or not
* Modify `rz_project_save` to do compression according to the
  `compression` bool parameter
* Update elsewhere (tests, rizin.c) according to the new function signature

DMaroo committed:
Modify `prj.decompress` to be a bool pref

DMaroo committed:
Merge branch 'dev' into local-dev

DMaroo committed:
Merge branch 'dev' into local-dev

XVilka commented on librz/core/cconfig.c:
```diff
@@ -3146,6 +3146,7 @@ RZ_API int rz_core_config_init(RzCore *core) {
 
 	/* prj */
 	SETPREF("prj.file", "", "Path of the currently opened project");
+	SETBPREF("prj.decompress", "true", "Decompress the project file while saving");
```
I think it should be `prj.compress` instead, and `Compress the project file` instead

DMaroo committed:
Rename `prj.decompress` config option to `prj.compress`

XVilka commented on librz/core/cmd_project.c:
```diff
@@ -26,13 +26,14 @@ RZ_IPI RzCmdStatus rz_project_save_handler(RzCore *core, int argc, const char **
 		file = argv[1];
 	}
-	RzProjectErr err = rz_project_save_file(core, file);
+	bool decompress = rz_config_get_b(core->config, "prj.compress");
```
Should be `compress` as well then.

XVilka replied:
See also other places with the same name problem.

DMaroo committed:
Merge branch 'dev' into local-dev

XVilka committed:
Fix

ret2libc commented on librz/core/project.c:
```diff
@@ -37,7 +39,20 @@ RZ_API RzProjectErr rz_project_save(RzCore *core, RzProject *prj, const char *fi
 
-RZ_API RzProjectErr rz_project_save_file(RzCore *core, const char *file) {
+RZ_API RzProjectErr rz_project_save_file(RzCore *core, const char *file, bool compress) {
+
```
```suggestion
```

ret2libc commented on librz/core/project.c:
```diff
@@ -37,7 +39,20 @@ RZ_API RzProjectErr rz_project_save(RzCore *core, RzProject *prj, const char *fi
-RZ_API RzProjectErr rz_project_save_file(RzCore *core, const char *file) {
+RZ_API RzProjectErr rz_project_save_file(RzCore *core, const char *file, bool compress) {
+
+	char *tmp_file;
```
```suggestion
	char *tmp_file = NULL;
```

ret2libc commented on librz/core/project.c:
```diff
@@ -47,13 +62,32 @@ RZ_API RzProjectErr rz_project_save_file(RzCore *core, const char *file) {
-	if (err == RZ_PROJECT_ERR_SUCCESS) {
-		rz_config_set(core->config, "prj.file", file);
+
+	if (!compress) {
+		if (err == RZ_PROJECT_ERR_SUCCESS) {
+			rz_config_set(core->config, "prj.file", file);
+		}
+		return err;
 	}
+
+	if (err != RZ_PROJECT_ERR_SUCCESS) {
+		goto tmp_file_err;
+	}
+
+	if (!rz_file_deflate(tmp_file, file)) {
+		err = RZ_PROJECT_ERR_COMPRESSION_FAILED;
+		goto tmp_file_err;
+	}
+
+	rz_config_set(core->config, "prj.file", file);
```
```suggestion
	if (err != RZ_PROJECT_ERR_SUCCESS) {
		goto tmp_file_err;
	}
	
	if (compress && !rz_file_deflate(tmp_file, file)) {
		err = RZ_PROJECT_ERR_COMPRESSION_FAILED;
		goto tmp_file_err;
	}

	rz_config_set(core->config, "prj.file", file);
```
something like this.

DMaroo replied:
Thanks. This makes it much simpler.

ret2libc reviewed (commented):
I don't see particular problems with compressing by default. Actually, on a project of a chrome binary analysis it saved quite a lot (something like ~1.4GB vs less than 200MB).

DMaroo committed:
Apply suggestions from code review

* Remove newline
* Initialize `tmp_file` to NULL

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

DMaroo committed:
Simplify conditional execution for project file compression

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>
